### PR TITLE
fix(frontend): standardize resolving thinking controls

### PR DESCRIPTION
**Pull request by KrishnanPrash**
## Summary

A client can ask a model to think, or not to think, in four different ways. Model families then read four different chat-template keys to find that instruction. Dynamo resolved those controls in more than one place, so the answer a model received depended on which render path served the request.

This PR gives the decision one owner. `normalize_reasoning_template_args` runs once at the HTTP boundary, for every chat request, before any dispatch. It now writes the decision in all three template dialects, so a family that reads `thinking` gets the same answer as one that reads `enable_thinking`. Both custom chat processors forward that result instead of deriving their own.

## Why three keys

No single key works for every model. `Qwen` and `GLM` read `enable_thinking`. `Kimi` and `DeepSeek-V4` read `thinking`. `MiniMax-M3` reads `thinking_mode`. Before this PR, `reasoning_effort` wrote only `enable_thinking`, so it reached Qwen and GLM and said nothing to the other three. The SGLang chat processor patched that gap on its own path. The vLLM chat processor did not. That is how the two paths drifted apart.

## Precedence

Highest first. A rung only applies when no rung above it decided.

1. The root `thinking` param
2. `thinking` or `enable_thinking` in `chat_template_args`
3. `thinking_mode` in `chat_template_args`
4. `reasoning_effort`, top-level first, then the nested copy. A nested value must be a string; `null` counts as absent
5. `--dyn-default-thinking-mode`
6. The model or chat template's own default

`adaptive` is the one value that decides *not* to decide, and where it came from matters. As the root `thinking` param it is the highest control, so it clears any toggle the client sent. Inside `chat_template_args` it sits below the bools, so an explicit `thinking` or `enable_thinking` beside it wins and replaces it.

## Request in, template args out

Every row below is pinned by a unit test.

| Request | `chat_template_args` after normalization |
|---|---|
| nothing | unchanged |
| `reasoning_effort: high` | `thinking: true`, `enable_thinking: true`, `thinking_mode: enabled`, `reasoning_effort: high` |
| `reasoning_effort: none` | `thinking: false`, `enable_thinking: false`, `thinking_mode: disabled`, `reasoning_effort: none` |
| `thinking: {"type": "enabled"}` | `thinking: true`, `enable_thinking: true`, `thinking_mode: enabled` |
| `thinking: {"type": "disabled"}` | `thinking: false`, `enable_thinking: false`, `thinking_mode: disabled` |
| `thinking: {"type": "adaptive"}` | `thinking_mode: adaptive` only, any client toggle removed |
| `thinking: {"type": "adaptive"}` + `reasoning_effort: high` | `thinking_mode: adaptive`, `reasoning_effort: high`, no toggle |
| `chat_template_args: {enable_thinking: false}` + `reasoning_effort: high` | `thinking: false`, `enable_thinking: false`, `thinking_mode: disabled`, `reasoning_effort: high` |
| `chat_template_args: {thinking_mode: disabled}` + `reasoning_effort: high` | `thinking: false`, `enable_thinking: false`, `thinking_mode: disabled`, `reasoning_effort: high` |
| `chat_template_args: {reasoning_effort: none}` | `thinking: false`, `enable_thinking: false`, `thinking_mode: disabled`, `reasoning_effort: none` |
| `chat_template_args: {enable_thinking: false}` | `thinking: false`, `enable_thinking: false`, `thinking_mode: disabled` |
| `chat_template_args: {thinking: false, thinking_mode: adaptive}` | `thinking: false`, `enable_thinking: false`, `thinking_mode: disabled` |
| `chat_template_args: {reasoning_effort: null}` | unchanged |

A grade other than `none` means think. That includes `minimal`, which asks for less reasoning, not for none.

## Parity with vLLM and SGLang

| Behavior | This PR | Dynamo before | vLLM | SGLang |
|---|---|---|---|---|
| `reasoning_effort` sets a thinking toggle | Yes, all three keys | Only `enable_thinking` | No, passed to the template untouched | No, passed to the template untouched |
| Accepts `reasoning_effort: none` | Yes | Yes | No, `low`/`medium`/`high` only | No, `low`/`medium`/`high` only |
| Top-level vs nested `reasoning_effort` | Top-level wins | Nested ignored | Top-level wins | Nested wins, it is popped and promoted |
| Root `thinking` param | Yes | Yes | Not supported | Not supported |
| Where the decision is made | Frontend only | Frontend and the SGLang processor | Engine | Engine |

Neither engine turns `reasoning_effort` into a thinking decision, and neither accepts `none`. Dynamo does both, so one request field behaves the same on every backend. That is a deliberate deviation, and it is the reason the SGLang processor grew its own `none` handling in the first place. With the frontend now writing every dialect, that handling is no longer needed and has been removed.

The one place Dynamo does not match either engine is nested `reasoning_effort`: vLLM lets the top-level field win, SGLang lets the nested one win. This PR follows vLLM.

## What changed

- `chat_completions.rs`: one precedence chain replaces the previous partial derivation. It lowercases `thinking_mode` to match the case-insensitive readers downstream, it no longer returns early when only `chat_template_args` carries a control, and it derives a toggle from a grade only when that grade is a string.
- `sglang_prepost.py`: the processor's own derivation is deleted. It forwards what the frontend resolved, which is what the vLLM processor already did.
- `prepost.py`: stops injecting `reasoning_effort: None` when the request has no grade. The guard above it already protected the nested value, so this line only added a null.
- `_thinking_parity.py`: a shared case table, asserted by both backend suites, so the two paths cannot drift again without a test failing.
- `FRONTEND_CASES.md`: adds `FRONTEND.10` for thinking-control forwarding.

## Validation

- `cargo test -p dynamo-llm --lib protocols::openai::chat_completions`: 84 passed, 0 failed.
- `cargo fmt --all` and `cargo clippy -p dynamo-llm --lib` clean.
- vLLM 0.27.1, real venv: `test_vllm_processor_unit.py` 119 passed.
- SGLang 0.5.17, real venv: `test_sglang_processor_unit.py` 227 passed.
- Both suites run the shared parity table, so every case above is checked on both backends.

### CODE REVIEW

File: lib/llm/src/protocols/openai/chat_completions.rs [hunk 1]
```diff
@@ -147,24 +147,34 @@ impl NvCreateChatCompletionRequest {
             return Ok(());
         }
 
+        let is_adaptive = matches!(thinking_mode, Some(OpenAiThinkingMode::Adaptive));
```
[the remark this answers is not in the document]
Comment (KrishnanPrash) [reply lib/llm/src/protocols/openai/chat_completions.rs]: b2fe54682b - False positive. `matches!` binds nothing, so it borrows rather than moves; `thinking_mode` is still owned at the `if let` below. The commit compiles and `cargo test -p dynamo-llm --lib protocols::openai::chat_completions` passes 81/81.

File: components/src/dynamo/frontend/prepost.py [hunk 2]
```diff
@@ -137,7 +137,19 @@ def _prepare_request(
...
+    # reserved key like `documents` would collide.
+    if isinstance(request, dict):
+        recovered = request.get("chat_template_args") or {}
+        for key in ("thinking", "enable_thinking", "thinking_mode"):
+            if key in recovered:
+                chat_template_kwargs.setdefault(key, recovered[key])
+    reasoning_effort = request_for_sampling.reasoning_effort
+    if reasoning_effort is not None:
+        chat_template_kwargs["reasoning_effort"] = reasoning_effort
+        chat_template_kwargs.setdefault("enable_thinking", reasoning_effort != "none")
```
[the remark this answers is not in the document]
Comment (KrishnanPrash) [reply components/src/dynamo/frontend/prepost.py]: b2fe54682b - Obsolete. `prepost.py` is no longer in this PR. #11047 landed the vLLM-path handling this branch carried, so the file was dropped and the branch rebuilt on main.

[the remark this answers is not in the document]
Comment (KrishnanPrash) [reply components/src/dynamo/frontend/prepost.py]: b2fe54682b - Obsolete. `prepost.py` is no longer in this PR. #11047 landed the vLLM-path handling this branch carried, so the file was dropped and the branch rebuilt on main.

File: lib/llm/src/protocols/openai/chat_completions.rs [hunk 3]
```diff
@@ -173,8 +183,11 @@ impl NvCreateChatCompletionRequest {
             }
         }
         if let Some(effort) = reasoning_effort {
-            args.entry("enable_thinking".to_string())
-                .or_insert_with(|| serde_json::Value::Bool(effort.as_str() != Some("none")));
+            // `adaptive` defers to the model, so a graded effort derives no toggle.
+            if !is_adaptive {
+                args.entry("enable_thinking".to_string())
```
[the remark this answers is not in the document]
Comment (KrishnanPrash) [reply lib/llm/src/protocols/openai/chat_completions.rs]: b2fe54682b - Real, but pre-existing on main and out of scope here. The `entry()` default protects only `enable_thinking`, not `thinking` or `thinking_mode`, and this PR did not change that: it only added the `!is_adaptive` guard. Widening it turns an adaptive-only fix into a precedence rewrite, so it belongs in its own change.

File: components/src/dynamo/frontend/prepost.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (KrishnanPrash) [reply components/src/dynamo/frontend/prepost.py]: b2fe54682b - Obsolete for the same reason as the other `prepost.py` comments: the file is no longer in this PR.

File: lib/llm/src/protocols/openai/chat_completions.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (KrishnanPrash) [reply lib/llm/src/protocols/openai/chat_completions.rs]: e730ccaafe - Correction: my earlier reply is stale. This is now fixed in this PR, not deferred. `reasoning_effort` no longer derives a toggle independently; the resolution is one precedence chain (client bool, then `thinking_mode`, then the grade) and whatever wins is written to all three keys. Both inputs you named are now pinned: `{"thinking": false}` + `reasoning_effort: "high"` and `{"thinking_mode": "disabled"}` + `reasoning_effort: "high"` each yield `enable_thinking: false`, in `test_explicit_thinking_bool_overrides_reasoning_effort` and `test_thinking_mode_expands_into_bool_dialects`.

File: lib/llm/src/protocols/openai/chat_completions.rs [hunk 4]
```diff
@@ -128,58 +128,79 @@ pub struct NvCreateChatCompletionRequest {
...
                     );
                 }
             }
         }
+        // Downstream readers compare modes case-insensitively; match them.
+        let mode = args
+            .get("thinking_mode")
+            .and_then(|v| v.as_str())
+            .map(str::to_ascii_lowercase);
+        if mode.as_deref() != Some("adaptive") {
```
Comment (indrajit96) [inline lib/llm/src/protocols/openai/chat_completions.rs]: Honor boolean precedence over nested adaptive mode

The documented precedence places thinking/enable_thinking above nested thinking_mode, but the adaptive short-circuit currently bypasses that precedence.

  For example:
```
  {
    "chat_template_args": {
      "thinking": false,
      "thinking_mode": "adaptive"
    }
  }
```
  When mode == "adaptive", resolution is skipped, leaving both values in the request. MiniMax consumes thinking_mode and therefore remains adaptive even though the higher-precedence boolean says disabled.

Suggested Fix: 

Please distinguish root-level adaptive from nested adaptive:

  - Root thinking: {"type":"adaptive"} remains highest precedence and should clear boolean toggles.
  - For controls inside chat_template_args, resolve thinking/enable_thinking first. If a boolean wins, overwrite or remove the lower-precedence adaptive mode.
  - Alternatively, reject contradictory nested controls with a 400.

Can we also add a Rust test for this exact combination.

File: lib/llm/src/protocols/openai/chat_completions.rs [hunk 5]
```diff
@@ -128,58 +128,79 @@ pub struct NvCreateChatCompletionRequest {
...
+                .find_map(|key| args.get(*key).and_then(|v| v.as_bool()))
+                .or(match mode.as_deref() {
+                    Some("enabled") => Some(true),
+                    Some("disabled") => Some(false),
+                    _ => None,
+                })
+                .or_else(|| {
+                    reasoning_effort
+                        .as_ref()
+                        .map(|effort| effort.as_str() != Some("none"))
```
Comment (indrajit96) [inline lib/llm/src/protocols/openai/chat_completions.rs]: Do not interpret null or non-string nested reasoning effort as enabled

Nested reasoning_effort is copied as an arbitrary JSON value. The decision then uses:
```
  effort.as_str() != Some("none")
```
  For JSON null, a number, or an object, as_str() returns None, making the comparison true. Therefore:
```
  {
    "chat_template_args": {
      "reasoning_effort": null
    }
  }
```
unexpectedly enables thinking and may also suppress the deployment default.
Suggested Fix : 

Can we derive a toggle only from an actual string:

```
  reasoning_effort
      .as_ref()
      .and_then(serde_json::Value::as_str)
      .map(|effort| effort != "none")
```
 JSON null should be treated as absent—or rejected explicitly—and must not produce a thinking toggle. Please add tests for nested null and one non-string value.

**Review by indrajit96 — APPROVED** [review]
LGTM!
With a couple of corner cases found please fix them, conditianlly approving!
Thanks a lot for adding this @KrishnanPrash

File: lib/llm/src/protocols/openai/chat_completions.rs (hunk 4 above)
Comment (KrishnanPrash) [reply lib/llm/src/protocols/openai/chat_completions.rs]: 9911e48e8d - Good catch, fixed. You are right that the short-circuit bypassed the documented ladder. The adaptive check moved from an outer guard into the chain, so the bool is consulted first and a nested `adaptive` only wins when no bool is present. Root `thinking: {"type":"adaptive"}` keeps its higher precedence for free, because that arm already removes both bools before this runs, so any bool reaching the chain is the client's own nested one. Your exact input now resolves to `thinking:false, enable_thinking:false, thinking_mode:disabled`, and it is pinned by `test_nested_adaptive_yields_to_client_bool`. I wrote that test before the fix and confirmed it failed on the old code first.

File: lib/llm/src/protocols/openai/chat_completions.rs (hunk 5 above)
Comment (KrishnanPrash) [reply lib/llm/src/protocols/openai/chat_completions.rs]: 9911e48e8d - Fixed, and it was worse than the toggle alone. `null` also counted as a present grade, which suppressed the deployment default in `apply_default_thinking_mode_to_template_kwargs`, since `reasoning_effort` is one of its control keys. Two changes: the nested lookup now filters `null` so it is treated as absent, and the grade derivation uses `as_str()` so a non-string decides nothing rather than reading as not-"'none"'. Nested `null` now leaves the request untouched, and a number forwards as sent without deriving a toggle. Pinned by `test_nested_reasoning_effort_decides_only_when_a_string`, which covers both values and also failed on the old code first.

**Review by dagil-nvidia — APPROVED** [review]

**Review by athreesh — APPROVED** [review]
